src/cargo/util/command_prelude.rs

```diff
@@ -986,6 +989,49 @@ pub fn root_manifest(manifest_path: Option<&Path>, gctx: &GlobalContext) -> Carg
     }
 }
 
+pub fn get_targets_from_rustup() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustup")
+        .arg("target")
+        .arg("list")
+        .arg("--installed")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
+}
+
+pub fn get_target_from_rustc() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustc")
```

Comment:
We should probably be using https://docs.rs/cargo/latest/cargo/util/context/struct.GlobalContext.html#method.load_global_rustc for this

I wonder if we should do something to allow caching of this.    To do it properly, we'd need to load a `Workspace` which can be a bit expensive, possibly making this slower than just calling rustc every time.  I wonder if we could instead have an alternative form of this for completions.